File: retrocpu_boot_monitor/mn1613/test/handshake/handshake_chg_mode_test.ts
/**
 * g_bios_mode_set（CPU→IO コマンド 10h）
 * 根拠: HandShake.mdc「モード設定」/ boot_monitor.mdc / test_framework.mdc
 */
import {
  createSessionFromSettings,
  expect,
  test,
  type IoBoardHandshakeMock,
  type Mn1613AsmSession,
} from "../../../../retrocpu_test_framework/src/index.js";
import {
  mn1613MonHandshakeSettings,
  withMn1613CpuLog,
} from "../mn1613_mon_settings.js";

const BASE_REGS = {
  R2: 0x2222,
  R3: 0x3333,
  R4: 0x4444,
} as const;

/** モニターモード（ハンドシェイク 10h） */
const MODE_MONITOR = 0;
/** フリーモード（ハンドシェイク 10h） */
const MODE_FREE = 1;

const session: Mn1613AsmSession = createSessionFromSettings(
  withMn1613CpuLog(mn1613MonHandshakeSettings, import.meta.url),
);

/**
 * g_main 済み＋ ioMock handshake で 1 ケースを実行する。
 * @param fn 本体（session / mock 利用可）
 */
async function withCase(
  fn: (s: Mn1613AsmSession, mock: IoBoardHandshakeMock) => Promise<void>,
): Promise<void> {
  session.reload();
  try {
    await session.runInit();
    await fn(session, session.requireHandshakeMock());
  } finally {
    await session.detachIoMock();
  }
}

/**
 * g_bios_mode_set を呼び、CPU→IO 1 トランザクションと並行する。
 * @param mock IO モック
 * @param mode 0=モニター / 1=フリー（それ以外は IO が NG）
 */
async function callModeSet(
  mock: IoBoardHandshakeMock,
  mode: number,
): Promise<void> {
  await Promise.all([
    session.call("g_bios_mode_set", {
      registers: { ...BASE_REGS, R0: mode },
    }),
    mock.handleOneRequest(),
  ]);
}

test("フリーモードを設定すると IO の mode が 1 になる", async () => {
  await withCase(async (s, mock) => {
    expect(mock.state.mode).toBe(MODE_MONITOR);
    await callModeSet(mock, MODE_FREE);
    s.expectRegisters({ R0: 0 });
    expect(mock.state.mode).toBe(MODE_FREE);
  });
});

test("モニターモードを設定すると IO の mode が 0 になる", async () => {
  await withCase(async (s, mock) => {
    mock.state.mode = MODE_FREE;
    await callModeSet(mock, MODE_MONITOR);
    s.expectRegisters({ R0: 0 });
    expect(mock.state.mode).toBe(MODE_MONITOR);
  });
});

test("不正なモードは NG で、IO の mode は変わらない", async () => {
  await withCase(async (s, mock) => {
    expect(mock.state.mode).toBe(MODE_MONITOR);
    await callModeSet(mock, 2);
    s.expectRegisters({ R0: 1 });
    expect(mock.state.mode).toBe(MODE_MONITOR);
  });
});

test("R3/R4 は呼び出しの前後で保たれる", async () => {
  await withCase(async (s, mock) => {
    await callModeSet(mock, MODE_FREE);
    s.expectRegisters({ R0: 0, R3: BASE_REGS.R3, R4: BASE_REGS.R4 });
  });
});

File: retrocpu_boot_monitor/test/mn1613/handshake/handshake_chg_mode_test.rs
use std::sync::Arc;

use retrocpu_test_framework_rs::{
    create_session_from_settings, CallOptions, CallRegisters, CodeTestIoMockEntry, FrameworkError,
    IoBoardHandshakeMock, JsonTestSettings, Mn1613AsmSession,
};

const MODE_MONITOR: u16 = 0;
const MODE_FREE: u16 = 1;

fn base_regs() -> CallRegisters {
    CallRegisters {
        r2: Some(0x2222),
        r3: Some(0x3333),
        r4: Some(0x4444),
        ..Default::default()
    }
}

fn handshake_settings() -> JsonTestSettings {
    let mut s = super::mn1613_rs_settings();
    s.io_mock = Some(vec![CodeTestIoMockEntry::Handshake]);
    s
}

fn with_case<F>(f: F) -> Result<(), FrameworkError>
where
    F: FnOnce(&mut Mn1613AsmSession, Arc<IoBoardHandshakeMock>) -> Result<(), FrameworkError>,
{
    let mut s = create_session_from_settings(&handshake_settings(), None)?;
    s.run_init()?;
    let mock = s.require_handshake_mock()?;
    let ret = f(&mut s, Arc::clone(&mock));
    s.detach_io_mock();
    ret
}

fn call_mode_set(
    s: &mut Mn1613AsmSession,
    mock: &Arc<IoBoardHandshakeMock>,
    mode: u16,
) -> Result<(), FrameworkError> {
    mock.run_with_cpu_to_io_request(|| {
        let _ = s.call(
            "g_bios_mode_set_",
            CallOptions {
                registers: Some(CallRegisters {
                    r0: Some(mode),
                    ..base_regs()
                }),
                ..Default::default()
            },
        )?;
        Ok(())
    })?;
    Ok(())
}

#[test]
fn set_mode_free_updates_io_state() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        assert_eq!(
            mock.state.lock().expect("state lock").mode,
            MODE_MONITOR as u8
        );
        call_mode_set(s, &mock, MODE_FREE)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )?;
        assert_eq!(mock.state.lock().expect("state lock").mode, MODE_FREE as u8);
        Ok(())
    })
}

#[test]
fn set_mode_monitor_updates_io_state() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        mock.state.lock().expect("state lock").mode = MODE_FREE as u8;
        call_mode_set(s, &mock, MODE_MONITOR)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                ..Default::default()
            },
            None,
        )?;
        assert_eq!(
            mock.state.lock().expect("state lock").mode,
            MODE_MONITOR as u8
        );
        Ok(())
    })
}

#[test]
fn invalid_mode_returns_ng_and_keeps_mode() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        assert_eq!(
            mock.state.lock().expect("state lock").mode,
            MODE_MONITOR as u8
        );
        call_mode_set(s, &mock, 2)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(1),
                ..Default::default()
            },
            None,
        )?;
        assert_eq!(
            mock.state.lock().expect("state lock").mode,
            MODE_MONITOR as u8
        );
        Ok(())
    })
}

#[test]
fn mode_set_preserves_r3_r4() -> Result<(), FrameworkError> {
    with_case(|s, mock| {
        call_mode_set(s, &mock, MODE_FREE)?;
        s.expect_registers(
            &CallRegisters {
                r0: Some(0),
                r3: Some(0x3333),
                r4: Some(0x4444),
                ..Default::default()
            },
            None,
        )
    })
}
